### ISSUE TITLE
RWKV-beta pre-compiled kernel on windows

### ISSUE THREAD

**Issue by LXY1226** [josStorer/RWKV-Runner#180]
_此Issue的作用讨论大于汇报_
_This is a discussion issue_

我想尝试试用Runner中的CUDA-Beta(更快)，但在编译上面踩了一堆坑，放在这里记录一下(?)
预编译的kernel二进制在最下面。

Environment:

| Hardware | Model |
| -- | -- |
| CPU | AMD Ryzen 5 5600H with Radeon Graphics |
| MEM | 16GB DDR4 |
| GPU | NVIDIA GeForce RTX 3050 Laptop (4GB VRAM) |

| Software | Version |
| -- | -- |
| Windows | Win11 Pro 22621.2134 |
| NVIDIA Driver | 31.0.15.3623(2023/6/8) |
| CUDA DevKit | cuda_11.7.1_windows_network.exe |
| Python (External installed) | Python 3.10.11 |
| VS Build Tool | VS 2019 Build Tool |

#### I recommend you read full instructions then perform it, some QTE needed
Step for compiling this stub in windows:
- Install them all
- Install packages into system interpreter:
``` 
python.exe -m pip install torch==1.13.1 torchvision==0.14.1 torchaudio==0.13.1 --index-url https://download.pytorch.org/whl/cu117
python -m pip install -r ./backend-python/requirements.txt -i https://pypi.tuna.tsinghua.edu.cn/simple
```
- double-click RWKV-Runner.exe and it will extract `...\backend-python\rwkv_pip\beta\cuda` and it's actually the kernel need to compile.
- Patch `...\backend-python\rwkv_pip\beta\cuda\model.py:L89`:
```
    if LoadPreCompileLibrary("wkv_cuda") is False:
        from torch.utils.cpp_extension import load

        load(
            name=f"wkv_cuda",
            sources=[
                f"{current_path}/cuda/wrapper.cpp",
                f"{current_path}/cuda/operators.cu",
                f"{current_path}/cuda/gemm_fp16_cublas.cpp",
                f"{current_path}/cuda/att_one.cu",
                f"{current_path}/cuda/att_seq.cu",
                f"{current_path}/cuda/ffn.cu",
            ],
>>>>       extra_ldflags=["cublas.lib"], # Add this line
            verbose=True,
            extra_cflags=[
                "/std=c++17",
            ],
            is_python_module=False,
        )
```
- Open a `x64 native tools command prompt for VS 2019` (you can find it in your `start menu > all programs`)
- and cd into `...\backend-python\rwkv_pip\beta\cuda` 
```
set RWKV_JIT_ON=1
set RWKV_CUDA_ON=1
python model.py
```
- QTE Here: YOU SHOULD RECORD the firstline for path used for torch extenions from output like:
```
Using **C:\Users\X\AppData\Local\torch_extensions\torch_extensions\Cache\py310_cu117** as PyTorch extensions root...
```
- and your pyd extension is ready in `C:\Users\X\AppData\Local\torch_extensions\torch_extensions\Cache\py310_cu117`
- copy it into `...\backend-python\rwkv_pip\beta\wkv_cuda.cp310-win_amd64.pyd`
- and it should works?
- If you received notice like `'str' object has no attribute 'is_meta'`
- Comment out `...\backend-python\rwkv_pip\beta\model.py:219-224`
- and it works again

当前的RWKV-Runner拉取打包好的、用于嵌入的python，所以可以预先编译的cuda kernel二进制。
Since RWKV-Runner fetchs zipped embed version of python, compilation of kernel binarys in earlier is possiable.

Safety notice:
### It's super unsafe for running untrusted executables in your machine

但下载那个zip然后解压到`...\backend-python\rwkv_pip\beta`就能用Beta kernel了
But just install cuda runtime(11) and then place this .pyd into `backend-python\rwkv_pip\beta`

[wkv_cuda.cp310-win_amd64.pyd.zip](https://github.com/josStorer/RWKV-Runner/files/12510077/wkv_cuda.cp310-win_amd64.pyd.zip)

sha1sum of **`wkv_cuda.cp310-win_amd64.pyd`** `76a9ee4a580c8360134ec704533d870acd2e9eac`

**Comment by josStorer** [comment]
Included in [v1.4.7](https://github.com/josStorer/RWKV-Runner/releases/tag/v1.4.7), recompiled with beta 0.8.5

**Comment by josStorer** [comment]
The latest version has improved cuda kernel compatibility (compute compatibility 5.3, Jetson Nano, MX Series, Nvidia 10 Series+)
